src/node_dir.cc

```diff
@@ -0,0 +1,351 @@
+  dir->closed_ = true;
+
+  FSReqBase* req_wrap_async = GetReqWrap(env, args[0]);
+  if (req_wrap_async != nullptr) {  // close(req)
+    AsyncCall(env, req_wrap_async, args, "closedir", UTF8, AfterClose,
+              uv_fs_closedir, dir->dir());
+  } else {  // close(undefined, ctx)
+    CHECK_EQ(argc, 2);
+    FSReqWrapSync req_wrap_sync;
+    FS_DIR_SYNC_TRACE_BEGIN(closedir);
+    SyncCall(env, args[1], &req_wrap_sync, "closedir", uv_fs_closedir,
+             dir->dir());
+    FS_DIR_SYNC_TRACE_END(closedir);
+  }
+}
+
+void AfterDirReadSingle(uv_fs_t* req) {
+  FSReqBase* req_wrap = FSReqBase::from_req(req);
+  FSReqAfterScope after(req_wrap, req);
+
+  if (!after.Proceed()) {
+    return;
+  }
+
+  Environment* env = req_wrap->env();
+  Isolate* isolate = env->isolate();
+  Local<Value> error;
+
+  if (req->result == 0) {
+    // Done
+    Local<Value> done = Null(isolate);
+    req_wrap->Resolve(done);
+    return;
+  }
+
+  uv_dir_t* dir = static_cast<uv_dir_t*>(req->ptr);
+  req->ptr = nullptr;
+
+  // Single entries are returned without an array wrapper
+  uv_dirent_t ent = dir->dirents[0];
+
+  MaybeLocal<Value> filename =
+    StringBytes::Encode(isolate,
+                        ent.name,
+                        req_wrap->encoding(),
+                        &error);
+  if (filename.IsEmpty())
+    return req_wrap->Reject(error);
+
+
+  Local<Array> result = Array::New(isolate, 2);
+  result->Set(env->context(),
+              0,
+              filename.ToLocalChecked()).FromJust();
+  result->Set(env->context(),
+              1,
+              Integer::New(isolate, ent.type)).FromJust();
+  req_wrap->Resolve(result);
+}
+
+
+void DirHandle::Read(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  const int argc = args.Length();
+  CHECK_GE(argc, 2);  // encoding, req?, ctx?
+
+  const enum encoding encoding = ParseEncoding(isolate, args[0], UTF8);
+
+  DirHandle* dir;
+  ASSIGN_OR_RETURN_UNWRAP(&dir, args.Holder());
+
+  FSReqBase* req_wrap_async = static_cast<FSReqBase*>(GetReqWrap(env, args[1]));
+  if (req_wrap_async != nullptr) {  // dir.read(encoding, req)
+    AsyncCall(env, req_wrap_async, args, "readdir", encoding,
+              AfterDirReadSingle, uv_fs_readdir, dir->dir());
+  } else {  // dir.read(encoding, undefined, ctx)
+    CHECK_EQ(argc, 3);
+    FSReqWrapSync req_wrap_sync;
+    FS_DIR_SYNC_TRACE_BEGIN(readdir);
+    int err = SyncCall(env, args[2], &req_wrap_sync, "readdir", uv_fs_readdir,
+                       dir->dir());
+    FS_DIR_SYNC_TRACE_END(readdir);
+    if (err < 0) {
+      return;  // syscall failed, no need to continue, error info is in ctx
+    }
+
+    if (req_wrap_sync.req.result == 0) {
+      // Done
+      Local<Value> done = Null(isolate);
+      args.GetReturnValue().Set(done);
+      return;
+    }
+
+    CHECK_GE(req_wrap_sync.req.result, 0);
+    uv_dirent_t ent = dir->dir()->dirents[0];
```

Comment:
```suggestion
    const uv_dirent_t& ent = dir->dir()->dirents[0];
```

(To avoid copying the struct when we don’t have to)